### PR TITLE
feat: add support for google-cloud-pubsub closes #25

### LINKED ISSUE #25: Add GCP Cloud PubSub handler

### PR DESCRIPTION
**Pull request by Mohsin-Ul-Islam**

### CODE REVIEW

**Comment by melvinkcx** [comment]
@Mohsin-Ul-Islam Hi, thank you for your contribution =) Would you please also include test cases to cover the usage of the handler?

**Comment by Mohsin-Ul-Islam** [comment]
@melvinkcx Hi, I have added a test for GCP pubsub handler. I couldn't find a mock library for google cloud pubsub python client, so I am testing it with google cloud pubsub emulator. More information on emulator can be found here: https://cloud.google.com/pubsub/docs/emulator

In order to spin up and tear down docker container I am using testcontainers library to spin up our pubsub emulator. I am a beginner in open source, so any guidance would be appreciated. Thanks a lot!

**Comment by melvinkcx** [comment]
@Mohsin-Ul-Islam I took a look at your test cases. The approach you take (integration tests) will significantly inflate the dependencies required to run the tests and for development (Docker, etc). I would suggest you mock the google clients yourself, and verify that the methods (`self._client.publish(self._topic_path, self.format_message(event))`, etc) are invoked correctly instead.

**Comment by melvinkcx** [comment]
The tests are failing, you might need to update the "Install dependencies" step of the github workflow `.github/workflows/tests.yml` as well

**Comment by Mohsin-Ul-Islam** [comment]
Sure, I will mock the google cloud pubsub client.

**Comment by melvinkcx** [comment]
Hi, @Mohsin-Ul-Islam, it seems like the test is now fixed. Would you please confirm if the PR is ready for review? =)

**Comment by Mohsin-Ul-Islam** [comment]
@melvinkcx yes the PR is ready for review. Thanks!

File: fastapi_events/handlers/gcp.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+from fastapi_events.errors import ConfigurationError
+from fastapi_events.handlers.base import BaseEventHandler
+from fastapi_events.typing import Event
+
+
+def _json_serializer(event: Event) -> str:
+    return json.dumps(event, default=str)
+
+
+class GoogleCloudPubSubHandler(BaseEventHandler):
```
Comment (melvinkcx) [inline fastapi_events/handlers/gcp.py]: @Mohsin-Ul-Islam I believe there will be needs to forward events to multiple topics depending on the event name, as some folks using AWS SNS (the pub/sub offering by AWS) have indicated.

What about renaming `GoogleCloudPubSubHandler` to `GoogleCloudSimplePubSubHandler` to allow room in the future to implement a GCP pub/sub handler with more advanced routing capabilities?

 I think it'd be better if we rename it to something such as `GoogleCloudSimplePubSubHandler` to indicate that the proces

File: fastapi_events/handlers/gcp.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+        if serializer is not None and not callable(serializer):
+            raise ConfigurationError("serializer must be of type Callable")
+
+        self._max_batch_size = max_batch_size
+
+        # Publish messages as soon as there are max_messages
+        # or 1 second is passed
+        self._batch_settings = pubsub_v1.types.BatchSettings(
+            max_messages=self._max_batch_size, max_latency=1
```
Comment (melvinkcx) [inline fastapi_events/handlers/gcp.py]: We should allow the users to customise their BatchSettings:

This is a suggestion:
```

    def __init__(
        self,
        project_id: str,
        topic_id: str,
        max_batch_size: int = 1000,
        batch_settings_kwargs: Dict[str, Any] = None,
        serializer: Callable[[Event], str] = None,
    ) -> None:
        ...
        self._batch_settings = pubsub_v1.types.BatchSettings(
            max_messages=self._max_batch_size, **batch_settings_kwargs
        )
```

**Review by melvinkcx — CHANGES_REQUESTED** [review]
Hi @Mohsin-Ul-Islam, thanks for contributing. Please check the comments I added. Also, please update the README to document the usage of your handler =)

File: fastapi_events/handlers/gcp.py (hunk 2 above)
Comment (Mohsin-Ul-Islam) [reply fastapi_events/handlers/gcp.py]: Hi, @melvinkcx, thank you so much for the review. The changes requested makes perfect sense. I will update the PR.

**Review by melvinkcx — APPROVED** [review]
Thank you for your patience. LGTM, approved!

**Pull request merged**